control_plane/src/pageserver.rs

```diff
@@ -304,42 +349,61 @@ impl PageServerNode {
         };
         io::stdout().flush().unwrap();
         match kill(pid, sig) {
-            Ok(_) => (),
+            Ok(()) => (),
             Err(Errno::ESRCH) => {
                 println!("Pageserver with pid {pid} does not exist, but a PID file was found");
                 return Ok(());
             }
-            Err(err) => bail!(
-                "Failed to send signal to pageserver with pid {pid}: {}",
-                err.desc()
-            ),
+            Err(e) => anyhow::bail!("Failed to send signal to pageserver with pid {pid}: {e}"),
         }
 
         // Wait until process is gone
-        for i in 0..600 {
-            let signal = None; // Send no signal, just get the error code
-            match kill(pid, signal) {
-                Ok(_) => (), // Process exists, keep waiting
-                Err(Errno::ESRCH) => {
-                    // Process not found, we're done
+        for _ in 0..RETRIES {
+            match self.has_node_stopped() {
+                Ok(true) => {
                     println!("done!");
                     return Ok(());
                 }
-                Err(err) => bail!(
-                    "Failed to send signal to pageserver with pid {}: {}",
-                    pid,
-                    err.desc()
-                ),
-            };
-
-            if i % 10 == 0 {
-                print!(".");
-                io::stdout().flush().unwrap();
+                Ok(false) => {
+                    print!(".");
+                    io::stdout().flush().unwrap();
+                    thread::sleep(Duration::from_secs(1))
+                }
+                Err(e) => {
+                    println!("Pageserver failed to stop: {e:#}");
+                    return Err(e);
+                }
             }
-            thread::sleep(Duration::from_millis(100));
         }
 
-        bail!("Failed to stop pageserver with pid {pid}");
+        anyhow::bail!("Failed to stop pageserver with pid {pid}");
+    }
+
+    fn has_node_stopped(&self) -> anyhow::Result<bool> {
+        match self.check_status() {
+            Ok(()) => Ok(false),
+            Err(_) => {
+                // node does not respond to status checks, ensure that the process does not run by
+                // checking the pid file for lock presence.
+                let pid_file_path = self.pid_file();
+                let pid_file = fs::OpenOptions::new()
+                    .read(true)
+                    .open(&pid_file_path)
+                    .with_context(|| {
+                        format!("Failed to open pid file {pid_file_path:?} for reading")
+                    })?;
+                match fcntl::flock(pid_file.as_raw_fd(), fcntl::FlockArg::LockExclusiveNonblock) {
```

Comment:
1. Both safekeeper and pageserver code have very similar code to handle node started/stopped checks.
They are slightly different in types though: Err is different and we have to mach over it to check things.
I have left the code duplication since it's only two places for now.

2. We have lost the ability to check via `kill(pid, None)` on the background task: apparently, `subprocess` `detach` logic on the subprocesses attaches them to `PID 1` process as a parent after we've exited the code that spawns it. 
Apparently that never allows us to get `Err(Errno::ESRCH)` on such call.

Instead now, I check that the lockfile lock is released instead.

Reply:
Not sure second paragraph is correct. 

`man 2 kill` says:

```
For a process to have permission to send a signal, it must either be privileged 
(under Linux: have the CAP_KILL capability in the user namespace of the 
target process), or the real or effective user ID of the sending process must 
equal the real or saved set-user-ID of the target  process.
```

And user ids for both processes should be the same. 

Tested with this patch on top of your PR and it worked for me:
```
diff --git a/control_plane/src/pageserver.rs b/control_plane/src/pageserver.rs
index 9e4ecbce..61b9e80d 100644
--- a/control_plane/src/pageserver.rs
+++ b/control_plane/src/pageserver.rs
@@ -359,7 +359,7 @@ impl PageServerNode {
 
         // Wait until process is gone
         for _ in 0..RETRIES {
-            match self.has_node_stopped() {
+            match self.has_node_stopped(pid) {
                 Ok(true) => {
                     println!("done!");
                     return Ok(());
@@ -379,7 +379,21 @@ impl PageServerNode {
         anyhow::bail!("Failed to stop pageserver with pid {pid}");
     }
 
-    fn has_node_stopped(&self) -> anyhow::Result<bool> {
+    fn has_node_stopped(&self, pid: Pid) -> anyhow::Result<bool> {
+        match kill(pid, None) {
+            Ok(_) => false, // Process exists, keep waiting
+            Err(Errno::ESRCH) => {
+                // Process not found, we're done
+                println!("done!");
+                return Ok(true);
+            }
+            Err(err) => bail!(
+                "Failed to send signal to pageserver with pid {}: {}",
+                pid,
+                err.desc()
+            ),
+        };
+
         match self.check_status() {
             Ok(()) => Ok(false),
             Err(_) => {
diff --git a/pageserver/src/lib.rs b/pageserver/src/lib.rs
index 85a9ba17..3a8ed5c3 100644
--- a/pageserver/src/lib.rs
+++ b/pageserver/src/lib.rs
@@ -22,7 +22,7 @@ pub mod walreceiver;
 pub mod walrecord;
 pub mod walredo;
 
-use std::collections::HashMap;
+use std::{collections::HashMap, time::Duration};
 
 use tracing::info;
 use utils::id::{TenantId, TimelineId};
@@ -80,7 +80,7 @@ pub async fn shutdown_pageserver(exit_code: i32) {
 
     // There should be nothing left, but let's be sure
     task_mgr::shutdown_tasks(None, None, None).await;
-
+    std::thread::sleep(Duration::from_secs(10));
     info!("Shut down successfully completed");
     std::process::exit(exit_code);
 }

```

Reply:
> We have lost the ability to check via kill(pid, None) on the background task: apparently, subprocess detach logic on the subprocesses attaches them to PID 1 process as a parent after we've exited the code that spawns it.
Apparently that never allows us to get Err(Errno::ESRCH) on such call.

That can't be right. Calling `detach` only sets a flag on the `Popen` struct, and the flag is only checked on Drop. It cannot have any effect on how `kill()` behaves

Reply:
Thank you for the patch.
I think it works due to `+            Ok(_) => false, // Process exists, keep waiting` line not returning, but continuing into the same "check status and lockfile" code.

I've tried the following:
```diff
diff --git a/control_plane/src/lib.rs b/control_plane/src/lib.rs
index 1ea806f1a..e7795aecd 100644
--- a/control_plane/src/lib.rs
+++ b/control_plane/src/lib.rs
@@ -7,6 +7,9 @@
 // local installations.
 //
 use anyhow::{anyhow, bail, Context, Result};
+use nix::errno::Errno;
+use nix::sys::signal::kill;
+use nix::unistd::Pid;
 use std::fs;
 use std::path::Path;
 
@@ -61,3 +64,13 @@ fn fill_aws_secrets_vars(mut cmd: subprocess::Exec) -> subprocess::Exec {
     }
     cmd
 }
+
+fn is_process_stopped(pid: Pid) -> anyhow::Result<bool> {
+    match kill(pid, None) {
+        // Process exists, keep waiting
+        Ok(_) => Ok(false),
+        // Process not found, we're done
+        Err(Errno::ESRCH) => Ok(true),
+        Err(err) => anyhow::bail!("Failed to send signal to process with pid {pid}: {err}"),
+    }
+}
diff --git a/control_plane/src/pageserver.rs b/control_plane/src/pageserver.rs
index 9e4ecbce2..699fc8086 100644
--- a/control_plane/src/pageserver.rs
+++ b/control_plane/src/pageserver.rs
@@ -2,14 +2,12 @@ use std::collections::HashMap;
 use std::fs::{self, File};
 use std::io::{BufReader, Write};
 use std::num::NonZeroU64;
-use std::os::unix::prelude::AsRawFd;
 use std::path::{Path, PathBuf};
 use std::time::Duration;
 use std::{io, result, thread};
 
 use anyhow::{bail, Context};
 use nix::errno::Errno;
-use nix::fcntl;
 use nix::sys::signal::{kill, Signal};
 use nix::unistd::Pid;
 use pageserver_api::models::{
@@ -28,7 +26,7 @@ use utils::{
 };
 
 use crate::local_env::LocalEnv;
-use crate::{fill_aws_secrets_vars, fill_rust_env_vars, read_pidfile};
+use crate::{fill_aws_secrets_vars, fill_rust_env_vars, is_process_stopped, read_pidfile};
 
 #[derive(Error, Debug)]
 pub enum PageserverHttpError {
@@ -359,7 +357,7 @@ impl PageServerNode {
 
         // Wait until process is gone
         for _ in 0..RETRIES {
-            match self.has_node_stopped() {
+            match is_process_stopped(pid) {
                 Ok(true) => {
                     println!("done!");
                     return Ok(());
@@ -379,33 +377,6 @@ impl PageServerNode {
         anyhow::bail!("Failed to stop pageserver with pid {pid}");
     }
 
-    fn has_node_stopped(&self) -> anyhow::Result<bool> {
-        match self.check_status() {
-            Ok(()) => Ok(false),
-            Err(_) => {
-                // node does not respond to status checks, ensure that the process does not run by
-                // checking the pid file for lock presence.
-                let pid_file_path = self.pid_file();
-                let pid_file = fs::OpenOptions::new()
-                    .read(true)
-                    .open(&pid_file_path)
-                    .with_context(|| {
-                        format!("Failed to open pid file {pid_file_path:?} for reading")
-                    })?;
-                match fcntl::flock(pid_file.as_raw_fd(), fcntl::FlockArg::LockExclusiveNonblock) {
-                    // successfully locked the file, meaning that the old process does not lock it anymore
-                    // return, dropping the open file and releasing the lock
-                    Ok(()) => Ok(true),
-                    // pageserver process still runs, holding the file lock
-                    Err(nix::errno::Errno::EAGAIN) => Ok(false),
-                    Err(e) => Err(anyhow::anyhow!(
-                        "Failed to check pid file {pid_file_path:?} for locks: {e}"
-                    )),
-                }
-            }
-        }
-    }
-
     pub fn page_server_psql(&self, sql: &str) -> Vec<postgres::SimpleQueryMessage> {
         let mut client = self.pg_connection_config.connect(NoTls).unwrap();
 
diff --git a/control_plane/src/safekeeper.rs b/control_plane/src/safekeeper.rs
index 646183351..d00719207 100644
--- a/control_plane/src/safekeeper.rs
+++ b/control_plane/src/safekeeper.rs
@@ -1,5 +1,4 @@
 use std::io::Write;
-use std::os::unix::prelude::AsRawFd;
 use std::path::PathBuf;
 use std::sync::Arc;
 use std::time::Duration;
@@ -7,7 +6,6 @@ use std::{fs, io, result, thread};
 
 use anyhow::Context;
 use nix::errno::Errno;
-use nix::fcntl;
 use nix::sys::signal::{kill, Signal};
 use nix::unistd::Pid;
 use postgres::Config;
@@ -18,7 +16,7 @@ use utils::{connstring::connection_address, http::error::HttpErrorBody, id::Node
 
 use crate::local_env::{LocalEnv, SafekeeperConf};
 use crate::pageserver::PageServerNode;
-use crate::{fill_aws_secrets_vars, fill_rust_env_vars, read_pidfile};
+use crate::{fill_aws_secrets_vars, fill_rust_env_vars, is_process_stopped, read_pidfile};
 
 #[derive(Error, Debug)]
 pub enum SafekeeperHttpError {
@@ -292,11 +290,8 @@ impl SafekeeperNode {
 
         // Wait until process is gone
         for _ in 0..RETRIES {
-            match self.has_node_stopped() {
-                Ok(true) => {
-                    println!("done!");
-                    return Ok(());
-                }
+            match is_process_stopped(pid) {
+                Ok(true) => return Ok(()),
                 Ok(false) => {
                     print!(".");
                     io::stdout().flush().unwrap();
@@ -312,33 +307,6 @@ impl SafekeeperNode {
         anyhow::bail!("Failed to stop safekeeper {id} with pid {pid}");
     }
 
-    fn has_node_stopped(&self) -> anyhow::Result<bool> {
-        match self.check_status() {
-            Ok(()) => Ok(false),
-            Err(_) => {
-                // node does not respond to status checks, ensure that the process does not run by
-                // checking the pid file for lock presence.
-                let pid_file_path = self.pid_file();
-                let pid_file = fs::OpenOptions::new()
-                    .read(true)
-                    .open(&pid_file_path)
-                    .with_context(|| {
-                        format!("Failed to open pid file {pid_file_path:?} for reading")
-                    })?;
-                match fcntl::flock(pid_file.as_raw_fd(), fcntl::FlockArg::LockExclusiveNonblock) {
-                    // successfully locked the file, meaning that the old process does not lock it anymore
-                    // return, dropping the open file and releasing the lock
-                    Ok(()) => Ok(true),
-                    // safekeeper process still runs, holding the file lock
-                    Err(nix::errno::Errno::EAGAIN) => Ok(false),
-                    Err(e) => Err(anyhow::anyhow!(
-                        "Failed to check pid file {pid_file_path:?} for locks: {e}"
-                    )),
-                }
-            }
-        }
-    }
-
     fn http_request<U: IntoUrl>(&self, method: Method, url: U) -> RequestBuilder {
         // TODO: authentication
         //if self.env.auth_type == AuthType::NeonJWT {
diff --git a/pageserver/src/lib.rs b/pageserver/src/lib.rs
index 85a9ba17b..ed0daf143 100644
--- a/pageserver/src/lib.rs
+++ b/pageserver/src/lib.rs
@@ -22,7 +22,7 @@ pub mod walreceiver;
 pub mod walrecord;
 pub mod walredo;
 
-use std::collections::HashMap;
+use std::{collections::HashMap, time::Duration};
 
 use tracing::info;
 use utils::id::{TenantId, TimelineId};
@@ -80,7 +80,7 @@ pub async fn shutdown_pageserver(exit_code: i32) {
 
     // There should be nothing left, but let's be sure
     task_mgr::shutdown_tasks(None, None, None).await;
-
+    std::thread::sleep(Duration::from_secs(4));
     info!("Shut down successfully completed");
     std::process::exit(exit_code);
 }
```
and that did not work

Reply:
On the updated branch, here's the patch that does not work for me:
```diff
diff --git a/control_plane/src/lib.rs b/control_plane/src/lib.rs
index 332e43b14..0ec760c87 100644
--- a/control_plane/src/lib.rs
+++ b/control_plane/src/lib.rs
@@ -200,7 +200,7 @@ where
 
     // Wait until process is gone
     for _ in 0..RETRIES {
-        match node_had_stopped(pid_file, &node_status_check) {
+        match node_had_stopped(pid, &node_status_check) {
             Ok(true) => {
                 println!("\n{node_name} node stopped");
                 return Ok(());
@@ -220,31 +220,21 @@ where
     anyhow::bail!("{node_name} node with pid {pid} failed to stop in {RETRIES} attempts");
 }
 
-fn node_had_stopped<F>(pid_file_path: &Path, node_status_check: F) -> anyhow::Result<bool>
+fn node_had_stopped<F>(pid: Pid, _node_status_check: F) -> anyhow::Result<bool>
 where
     F: Fn() -> anyhow::Result<()>,
 {
-    match node_status_check() {
-        Ok(()) => Ok(false),
-        Err(_) => {
-            // node does not respond to status checks, ensure that the process does not run by
-            // checking the pid file for lock presence.
-            let pid_file = fs::OpenOptions::new()
-                .read(true)
-                .open(&pid_file_path)
-                .with_context(|| {
-                    format!("Failed to open pid file {pid_file_path:?} for reading")
-                })?;
-            match fcntl::flock(pid_file.as_raw_fd(), fcntl::FlockArg::LockExclusiveNonblock) {
-                // successfully locked the file, meaning that the old process does not lock it anymore
-                // return, dropping the open file and releasing the lock
-                Ok(()) => Ok(true),
-                // background process still runs, holding the file lock
-                Err(nix::errno::Errno::EAGAIN) => Ok(false),
-                Err(e) => Err(anyhow::anyhow!(
-                    "Failed to check pid file {pid_file_path:?} for locks: {e}"
-                )),
-            }
+    match kill(pid, None) {
+        Ok(_) => Ok(false), // Process exists, keep waiting
+        Err(Errno::ESRCH) => {
+            // Process not found, we're done
+            println!("done!");
+            Ok(true)
         }
+        Err(err) => bail!(
+            "Failed to send signal to pageserver with pid {}: {}",
+            pid,
+            err.desc()
+        ),
     }
 }
```

Reply:
Yeah, there is a problem (took a while to go past `neon_local init`). Started pageserver becomes a zombie since cli process that spawned it dies. Because of that pageserver process is still reported as existing by the result of `kill` but is dead otherwise. Not sure that this can be easily solved by some workaround. Most promising approach is to get rid of daemonize and manage storage components in tests directly with python's subprocess module

Reply:
Latest commit fixes all related issues:
* `start` and then `stop` are working fine with `kill -0`
* `init` command that failed is now fixed to explicitly start and kill the subprocess